autowarefoundation/autoware
# [Feature] expand control points for pure_pursuit

## Status
**DEVELOPMENT**

## Description
Change /pure_pursuit to reach the end of waypoints. autowarefoundation/autoware_ai#272
In order to maintain the stability of the system near the endpoint of waypoints, a virtual control points was added forward from the endpoint of waypoints.
![image](https://user-images.githubusercontent.com/33311630/45206975-81bed680-b2c1-11e8-9f19-3d0fa57706fd.png)
Extended waypoints are indicated by blue dots.

## Todos

Before autowarefoundation/autoware#1235 work: the steering angle becomes unstable near the end point.
![image](https://user-images.githubusercontent.com/33311630/44992750-25169f80-afd3-11e8-8a6a-05ea9ec0d05b.png)

After autowarefoundation/autoware#1235 work: the steering angle is stable, but it stops before reaching the end point.
![image](https://user-images.githubusercontent.com/33311630/44992790-424b6e00-afd3-11e8-934d-1c3bb87f3ad0.png)

This work: the steering angle is stable, and can reach the end point.
![image](https://user-images.githubusercontent.com/33311630/44992832-6ad36800-afd3-11e8-9b28-614dd6b3cd49.png)

## Steps to Test or Reproduce
Prepare any environment and any waypoints.
Use wf_sim to run a route tracking simulation.
At that time, replanning_mode of waypoint_loader should be turned off so that speed 0 is not output before reaching termination.
If you want to visualize the extended line, please display / expanded_waypoints_marker with rviz.
